reactors/1D_packbed.ipynb

```diff
@@ -0,0 +1,518 @@
+    "\n",
+    "    # Set current thermodynamic state for the gas and surface phases\n",
+    "    # Note: use unnormalized mass fractions and site fractions to avoid over-constraining the system)\n",
+    "    gas.set_unnormalized_mass_fractions(Y)\n",
+    "    gas.TP = T, p\n",
+    "    surf.set_unnormalized_coverages(Z)\n",
+    "    surf.TP = T, p\n",
+    "\n",
+    "    # Calculate required variables based on the current state\n",
+    "    coverages = surf.coverages  # surface site coverages\n",
+    "    sdot_g = surf.get_net_production_rates(\n",
+    "        \"gas\"\n",
+    "    )  # heterogeneous production rate of gas species\n",
+    "    sdot_s = surf.get_net_production_rates(\n",
+    "        \"Ru_surface\"\n",
+    "    )  # heterogeneous production rate of surface species\n",
+    "    wdot_g = np.zeros(n_gas)\n",
+    "    cp = gas.cp_mass  # specific heat of the mixture\n",
+    "    hk_g = gas.partial_molar_enthalpies  # partial enthalpies of gas species\n",
+    "\n",
+    "    if n_gas_reactions > 0:\n",
+    "        wdot_g = gas.net_production_rates  # homogeneous production rate of gas species\n",
+    "    mu = gas.viscosity  # viscosity of the gas-phase\n",
+    "\n",
+    "    # Calculate density using equation of state\n",
+    "    rho = gas.density\n",
+    "\n",
+    "    # Calculate flux term through the membrane\n",
+    "    memsp_pres = (\n",
+    "        p * gas.X[membrane_sp_ind]\n",
+    "    )  # partial pressure of membrane-permeable species\n",
+    "    membrane_flux = (\n",
+    "        -permeance * (memsp_pres - p_sweep) / W_g[membrane_sp_ind]\n",
+    "    )  # negative sign indicates the flux going out\n",
+    "\n",
+    "    # Conservation of total-mass\n",
+    "    sum_continuity = As * np.sum(sdot_g * W_g) + phi * np.sum(\n",
+    "        wdot_g * W_g\n",
+    "    )  # temporary variable\n",
+    "    res[offset_rhou] = (\n",
+    "        drhoudz - sum_continuity - area2vol * membrane_flux * membrane_present\n",
+    "    )\n",
+    "\n",
+    "    # Conservation of gas-phase species\n",
+    "    for k in range(n_gas):\n",
+    "        res[offset_Y + k] = (\n",
+    "            dYdz[k]\n",
+    "            + (\n",
+    "                Y[k] * sum_continuity\n",
+    "                - phi * wdot_g[k] * W_g[k]\n",
+    "                - As * sdot_g[k] * W_g[k]\n",
+    "            )\n",
+    "            / rhou\n",
+    "        )\n",
+    "        if k == membrane_sp_ind:\n",
+    "            res[offset_Y + k] -= area2vol * membrane_flux * membrane_present\n",
+    "\n",
+    "    # Conservation of site fractions (algebraic contraints in this example)\n",
+    "    # for k in range(n_surf):\n",
+    "    res[offset_Z : offset_Z + n_surf] = sdot_s\n",
+    "\n",
+    "    # For the species with largest site coverage, use the constraint sum(Zk) = 1\n",
+    "    ind_large = np.argmax(coverages)\n",
+    "    res[offset_Z + ind_large] = 1 - np.sum(coverages)\n",
```

Comment:
Won't this include the largest species in the sum? So that `res[offset_Z + ind_large]` will be close to zero? Hopefully the sum of the coverages is already close to 1..., or do I misunderstand the code here (totally possible 😄 )

Reply:
Yes. That's right. We can just use the equation `\sdot  = 0` for all species, without applying the constraint.  But it may end up in generating error depending on the tolerances.  Therefore, the strict constraint is typically applied for the largest species.

Reply:
Right, what I'm saying is that `np.sum(coverages)` will include the largest species, so `1 - np.sum(coverages)` will be almost zero, and you'd be setting the coverages for the largest species to almost zero. I think it should be `1 - np.sum(coverages[all_except_ind_large])`. Is that a misunderstanding of the code?

Reply:
The equations for `ida `solver are written in the form `res = f(x) - 0`. We are solving the constraint equation `np.sum(coverages) = 1` here. So the residual function is `res = 1 - np.sum(coverages)`. This has to include all species coverages.

Reply:
Got it, thank you! It's probably worth adding a code comment here to summarize your last comment, since this is different than CVODES

Reply:
Sure. I added a comment explaining the logic.